Optimize tag count resolution via caching

nick-funk opened the pull request:
## What does this PR do?

Updates the logic for calculating the tag counts on a story to retrieve them from a cache of counts on the story document.

Also adds logic to generate the tag counts on the story document if they're not available (migrates existing stories).

## What changes to the GraphQL/Database Schema does this PR introduce?

No changes.

## Does this PR introduce any new environment variables or feature flags? 

No.

## If any indexes were added, were they added to `INDEXES.md`?

No new indexes.

## How do I test this PR?

- Go to a story that you have used before
- Load the page
- Open your MongoDB compass, VS Code Mongo tools, etc and find the story document
- Confirm that a new field for `tags` has shown up under `commentCounts` on the story document
- Check that the counts are correct for how many comments are featured, written by an admin, moderator, staff, etc
- Use a Q&A and a Ratings & Review story and confirm the tags are accurate there as well while reviewing, asking questions, answering, experts, etc... 
 
## How do we deploy this PR?

No special considerations to deploy.

nick-funk committed:
Testing cursor based tag count traversal

Seems to be improving things for very large stories

nick-funk committed:
Revert "Testing cursor based tag count traversal"

This reverts commit 39716b70275c4c0e548391079841d0a4ee1f377a.

nick-funk committed:
Update tag counts on related comment counts

nick-funk committed:
Fix issue where all tag counts were updated to admin count

nick-funk committed:
Initialize story tag counts on request, otherwise retrieve them

nick-funk committed:
Hook up merging tag counts to propagate counts up past stories

kabeaty commented on src/core/server/models/comment/comment.ts:
```diff
@@ -1043,6 +1047,52 @@ export async function retrieveStoryCommentTagCounts(
+    const tags = result.get(id);
+    if (!tags) {
+      return {
+        [GQLTAG.ADMIN]: 0,
```
What are your thoughts on defining this in one common place instead of repeating since it would be easy to leave a tag out somewhere?

nick-funk replied:
I noticed we do a similar pattern in `empty.ts` for merging from zero-ed counts, have replaced all of these with the convenience function I used there.

kabeaty commented on src/core/server/graph/mutators/Comments.ts:
```diff
@@ -206,6 +207,16 @@ export const Comments = (ctx: GraphContext) => ({
       );
     }
 
+    await updateTagCommentCounts(
```
Do we also need to have this update tag comment counts when unfeaturing comments?

nick-funk replied:
Yes, we do! I will fix!

nick-funk replied:
Is fixed!

kabeaty commented on src/core/server/models/comment/counts/counts.ts:
```diff
@@ -139,6 +163,27 @@ export function mergeCommentModerationQueueCount(
+  for (const tagSet of tags) {
+    merged.total += tagSet.total;
+
+    merged.tags.EXPERT += tagSet.tags.EXPERT;
```
Do we not need `ADMIN` here?

nick-funk replied:
Yup yup! Fixed, thank you!

kabeaty reviewed (changes requested):
Looks good overall! I left a few comments throughout.

nick-funk committed:
Add ADMIN ops to tag counts merge operations

nick-funk committed:
Remove duplication for creation of zeroed tag counts

nick-funk committed:
Decrement tag counts when a comment is un-featured

nick-funk committed:
Pull siteID from comment so feature button works in mod queues

nick-funk committed:
Specify tenantID when initializing story tag counts

nick-funk committed:
Create refreshStoryCounts mutation

Allows us to refresh the tag counts in an emergency if
a story is acting up and has improper counts.

kabeaty commented on src/core/server/graph/mutators/Comments.ts:
```diff
@@ -35,6 +36,7 @@ import {
   GQLUnfeatureCommentInput,
 } from "coral-server/graph/schema/__generated__/types";
 
+import { hasTag } from "coral-server/models/comment";
```
Can you please move this import up with the others from `coral-server`?

nick-funk replied:
And here I thought TS sort imports was magic... I have now found its kryptonite. Yep, I'll fix this up!

kabeaty commented on src/core/server/graph/mutators/Stories.ts:
```diff
@@ -40,6 +41,7 @@ import {
   GQLUpdateStorySettingsInput,
 } from "coral-server/graph/schema/__generated__/types";
 
+import { initializeCommentTagCountsForStory } from "coral-server/models/comment";
```
Can you please move this import up with the others from `coral-server`?

kabeaty reviewed (changes requested):
Thanks for the updates! When testing in Q&A mode, I am seeing that the `QUESTION` tag doesn't seem to be incrementing when I ask a question, only the `UNANSWERED` tag. And when I answer a question, the `UNANSWERED` tag isn't decrementing. Can you look into this?

nick-funk commented:
> Thanks for the updates! When testing in Q&A mode, I am seeing that the `QUESTION` tag doesn't seem to be incrementing when I ask a question, only the `UNANSWERED` tag. And when I answer a question, the `UNANSWERED` tag isn't decrementing. Can you look into this?

I believe I have a fix for the `UNANSWERED` increment and decrementing. Will be pushing up shortly.

As for the `QUESTION` not changing when you're in a Q&A stream, that would be because `QUESTION` is for Ratings & Reviews, and not for Q&A. I know, very confusing, but that mode also has a questions tab. I will re-test that as well to make sure it's working.

nick-funk committed:
Update the tag counts when we remove unanswered tag

This happens when an expert answers a question by replying
to it.

nick-funk committed:
Fix import ordering

nick-funk committed:
Fix count before/after state when marking answered questions

kabeaty reviewed (approved):
Looking great now, thanks for updating!

nick-funk commented:
> Looking great now, thanks for updating!

Awesome! Glad to hear it!

Also, I tested Ratings & Reviews, it looks to be working just fine since it doesn't add extra silly layers of add/remove tags in the code like Q&A does with its `UNANSWERED` tag.

This should be good to go now!

nick-funk commented:
Throwing a `HOLD MERGE` on this as it might be hotfixed into `main` after QA testing, or it may end up in `develop`. Just don't want it prematurely merged until @tessalt gives the okay.
